### ISSUE TITLE
nix build fails using ssh binary cache as trusted-user, but not as root

### ISSUE THREAD

**Issue by matrss** [NixOS/nix#5133]
**Describe the bug**

Ssh binary cache does work for root on the client machine, but does not work for a different trusted user on the client machine. Testing connectivity to the store (`nix store ping ...`) works in both cases.

**Steps To Reproduce**

I have setup a ssh binary cache on my server using the NixOS option `nix.sshServe.enable` and added my ssh key.
I can reach the server from my client machine using `nix store ping --store ssh://nix-ssh@example.com` and `ssh nix-ssh@example.com`.

Actually using the cache with e.g. `nix build self#pandoc --extra-substituters ssh://nix-ssh@example.com` **works** as root, but does **not work** as a different, trusted, user. In the latter case, `nix-daemon` logs a `Permission denied (publickey,keyboard-interactive)` error.

When I add another, passwordless, ssh key for `nix-ssh` on the server and add
```
Host bin-cache
        HostName example.com
        Port 22
        User nix-ssh
        # Prevent using ssh-agent or another keyfile, useful for testing
        IdentitiesOnly yes
        IdentityFile ~/.ssh/nix-ssh@example.com
```
to the ssh configuration of **root** on the client, I can suddenly use the cache (with `--extra-substituters ssh://bin-cache`) as a **different trusted user** on the client.

To me it seems like `nix build` as root uses ssh-agent private keys to connect to the store, while `nix build` as a different user ignores ssh-agent for some reason.

**Expected behavior**

The binary cache to be working in both cases, as the ssh connection does work for both root and non-root accounts.

**`nix-env --version` output**

nix-env (Nix) 2.4pre20210707_02dd6bb

**Additional context**

My usual ssh key is on a yubikey in smartcard mode, ssh-agent is actually gpg-agent  with enabled ssh-agent support.

**Comment by regnat** [comment]
> To me it seems like nix build as root uses ssh-agent private keys to connect to the store, while nix build as a different user ignores ssh-agent for some reason.

Yes, it does: Running `nix build` as root bypasses the daemon, meaning that the substitution will run in the context of your `nix build` command (which has access to all the environment needed to access the ssh-agent). Otoh, `nix build` as a non-root user will forward everything to the daemon, which doesn’t know about your running agent

**Comment by matrss** [comment]
That sounds like it would not be trivial to have `nix build` use ssh-agent as non-root, if that's even possible. Still, from a users point of view I think this is a very unintuitive thing to do. Also, the first note on https://nixos.org/manual/nix/stable/#ssec-ssh-substituter is somewhat misleading then.

Maybe we could just make this clearer in the documentation?

**Comment by hlolli** [comment]
I struggled for quite a while getting remote builders to work, it just wouldn't use my private key no matter how much I tried. After running "sudo nix build" it worked, I would never have guessed doing that if I didn't read this ticket. I think in osx's issue, it could be a matter of setting the plist for the daemon in such a way that it finds ssh auth lock.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
